[E2 - E3 - E3.1] - Cargo Service

rambodghx opened the pull request:
Hi @mohammadKarimi,

To implement the cargo-related use case has described earlier, this PR has been created.

Please note the following considerations:

- Validation has been ignored at this phase.
- The naming conventions used were sourced from Google. I apologize for any inaccuracies or unfamiliarity with this domain.

mohammadKarimi commented on Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Factory.Cargo.TransportationFactory;
+
+public class AirTransportationFactory : ITransportationFactory
```
where are you using this Factory ?
i didn't find it.

rambodghx replied:
Fixed.

mohammadKarimi commented on Factory.Cargo/CargoService.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        return transportationMode switch
+        {
+            TransportationMode.Air => new AirTransportation(),
```
Air must be single instance.

rambodghx replied:
Fixed.

mohammadKarimi commented on Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:
```diff
@@ -0,0 +1,15 @@
+    private ITransportation? _transportation;
+
+    public ITransportation Create()
+        => _transportation ??= new AirTransportation();
```
consider it is not thread safe, but for this case it is enough.

rambodghx replied:
Thanks @mohammadKarimi, I think in main implementation we should use atomic operations like Interlock.CompareExchange.
What is your idea?

mohammadKarimi reviewed (commented):
perfect one,

now move to E3.1

Imagine it is a library and we make it as a package. so users or customers, can't change your code, but they want to add a new behavior like Truck. so, how you can solve it.

rambodghx commented:
@mohammadKarimi please checkout my last commit **71b7714**.
I didn't use the factory instances and that was my mistake because, at first step I implemented the factory pattern and then added the **AbstractFactory** pattern so I just forgot make change on **CargoService**.
sorry about that I fixed this.

rambodghx commented:
@mohammadKarimi your new ordered task , E3.1 , has been implemented in latest commit 84716ac.
Note the following considerations :
- This use-case have been implemented without DI pattern in order to comply the KISS rule.
- Validations have been ignored in this phase
- Enum have been changed to **String** because there is no inheritance concept within Enum.

I look forward for your review.

mohammadKarimi commented on Factory.Cargo/CargoFactoryApi.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Factory.Cargo;
+
+public class CargoFactoryApi
```
why did you add API? 
it is  a bad naming selection.

rambodghx replied:
@mohammadKarimi
I was looking for some distinguish between package class and last practice class.
I have changed the name from Api to Service.
Please checkout [024ad3e].

mohammadKarimi commented on Factory.Cargo/CargoFactoryApi.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    private Dictionary<string, ITransportationFactory> _transportationFactories = new();
+
+    public void AddBaseFactories()
```
why you did not move it into constructor? 
you user must know to call this method before every usage.

rambodghx replied:
I have enabled user to figure out this feature by optional flag on Ctor.
Please checkout [fb9738f].

mohammadKarimi commented on Factory.Cargo/CargoFactoryApi.cs:
```diff
@@ -0,0 +1,36 @@
+    public void AddTransportationFactory(string mode, ITransportationFactory transportationFactory)
+        => _transportationFactories.Add(mode, transportationFactory);
+
+    public ITransportation Create(string mode, string? origin = null, string? destination = null)
```
if you have 10 factories and each of them needs different parameter. so how to solve it ? are you going to add 10 different null parameter?

rambodghx replied:
@mohammadKarimi 
I've developed new way to pass parameters dynamically.
`
public class CargoFactoryService
{
 public ITransportation Create(Dictionary<string,object> parameters);
}
`

so the use now can call factory in this way
` public LibraryUser()
    {
        _factoryService.AddTransportationFactory("Truck", new TruckTransportationFactory());

        var truckTransportation = _factoryService.Create("Truck", new Dictionary<string, object>
        {
            { "TruckType", "Trailer" },
        });
    }`

mohammadKarimi commented on Factory.Cargo/CargoFactoryApi.cs:
```diff
@@ -0,0 +1,36 @@
+            throw new NotImplementedException($"The factory {mode} didn't define.");
+        try
+        {
+            return transportationFactory.Create(origin, destination);
```
if you have 10 differ factory. how to call them? 

in this case you are calling in a try catch and then call another in line 34.

rambodghx replied:
@mohammadKarimi 
this issue addressed on previous review comment.

rambodghx reviewed (commented):
@mohammadKarimi 
Please checkout comments that I've left on your review including new commits.
Thanks.
